command-line/test/Command/Script/ValidationSpec.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+module Command.Script.ValidationSpec
+    ( spec
+    ) where
+
+import Prelude
+
+import Cardano.Address.Script
+    ( ValidationLevel (..) )
+import Cardano.Address.Script.Parser
+    ( ErrValidateScript (..), prettyErrValidateScript )
+import Data.String.Interpolate
+    ( iii )
+import Test.Hspec
+    ( Spec, SpecWith, it, shouldBe, shouldContain )
+import Test.Utils
+    ( cli, describeCmd )
+
+spec :: Spec
+spec = do
+    describeCmd [ "script", "validate"] $ do
+        specScriptValidated RequiredValidation
+            [iii|#{verKey4}|]
+
+        specScriptValidated RequiredValidation
+            [iii|at_least 2 [ #{verKeyH1}, #{verKeyH2}, #{verKeyH3} ]|]
+
+        specScriptValidated RequiredValidation
+            [iii|at_least 2 [ #{verKeyH1}, #{verKeyH2}, active_from 10, active_until 25]|]
+
+        specScriptValidated RequiredValidation
+            [iii|at_least 2 [ #{verKeyH1}, #{verKeyH2}, active_from 6, active_until 15]|]
+
+        specScriptValidated RequiredValidation
+            [iii|at_least 2 [ #{verKeyH1}, #{verKeyH2}, active_from 6, active_until 25]|]
+
+        specScriptValidated RequiredValidation
+            [iii|all []|]
+
+        specScriptNotValidated Malformed RequiredValidation
+            [iii|any [ #{verKeyH1}, #{verKeyH2}, active_from a]|]
+
+        specScriptNotValidated EmptyList RecommendedValidation
+            [iii|all []|]
+
+        specScriptValidated RequiredValidation
+            [iii|at_least 2 [ active_from 11, active_until 25]|]
+
+        specScriptValidated RequiredValidation
+            [iii|at_least 2 [ #{verKeyH1},  active_from 11, active_until 25]|]
+
+        specScriptNotValidated ListTooSmall RecommendedValidation
+            [iii|at_least 2 [ #{verKeyH1},  active_from 11, active_until 25]|]
+
+        specScriptValidated RequiredValidation
+            [iii|at_least 2 [ #{verKeyH1},  active_from 11, active_until 25, active_until 30]|]
+
+        specScriptNotValidated RedundantTimelocks RecommendedValidation
+            [iii|at_least 1 [ #{verKeyH1},  active_from 11, active_until 25, active_until 30]|]
+
+        specScriptValidated RequiredValidation
+            [iii|any [ #{verKeyH1}, #{verKeyH2}, #{verKeyH2}]|]
+
+        specScriptNotValidated DuplicateSignatures RecommendedValidation
+            [iii|any [ #{verKeyH1}, #{verKeyH2}, #{verKeyH2}]|]
+
+        specScriptValidated RequiredValidation
+            [iii|at_least 0 [ #{verKeyH1}, #{verKeyH2} ]|]
+
+        specScriptNotValidated MZero RecommendedValidation
+            [iii|at_least 0 [ #{verKeyH1}, #{verKeyH2} ]|]
+
```

Comment:
Some interesting cases I stumbled upon. I thought they will all validate as all are "well-formed"... So I guess `--required` does check some logic after all. Now I'm a bit more confused as of what is the difference between `--required` and `--recommended` :sweat_smile:  

```
cardano-address script validate --required "any []"
Not validated: The script is ill-formed and is not going to be accepted by ledger.

cardano-address script validate --required "at_least 1 []"
Not validated: The script is ill-formed and is not going to be accepted by ledger.

cardano-address script validate --required "at_least 4 [$(cat script.1.xvk), $(cat script.2.xvk)]"
Not validated: The script is ill-formed and is not going to be accepted by ledger.

cardano-address script validate --required "at_least 0 []"
Validated.

```

Reply:
that is what ledger accepts. So in summary in required we have 1. ledger compliant validation, 2. syntax validation 3 checking if verification keys are valid (by looking at length and prefix of bech32. And we have recommended validation. I hope now it is clear. And it is reflected in Error hierarchy. see head of this PR